arwen/cryptoapi/cryptoei.go

```diff
@@ -0,0 +1,266 @@
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	return 0
+}
+
+//export verifyBLS
+func verifyBLS(
+	context unsafe.Pointer,
+	keyOffset int32,
+	messageOffset int32,
+	messageLength int32,
+	sigOffset int32,
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.VerifyBLS
+	metering.UseGas(gasToUse)
+
+	key, err := runtime.MemLoad(keyOffset, blsPublicKeyLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	message, err := runtime.MemLoad(messageOffset, messageLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	sig, err := runtime.MemLoad(sigOffset, blsSignatureLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	invalidSigErr := crypto.VerifyBLS(key, message, sig)
+	if invalidSigErr != nil {
+		return 1
+	}
+
+	return 0
+}
+
+//export verifyEd25519
+func verifyEd25519(
+	context unsafe.Pointer,
+	keyOffset int32,
+	messageOffset int32,
+	messageLength int32,
+	sigOffset int32,
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.VerifyEd25519
+	metering.UseGas(gasToUse)
+
+	key, err := runtime.MemLoad(keyOffset, ed25519PublicKeyLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	message, err := runtime.MemLoad(messageOffset, messageLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	sig, err := runtime.MemLoad(sigOffset, ed25519SignatureLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	invalidSigErr := crypto.VerifyEd25519(key, message, sig)
+	if invalidSigErr != nil {
+		return 1
+	}
+
+	return 0
+}
+
+//export verifySecp256k1
+func verifySecp256k1(
+	context unsafe.Pointer,
+	keyOffset int32,
+	keyLength int32,
+	messageOffset int32,
+	messageLength int32,
+	sigOffset int32,
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.VerifySecp256k1
+	metering.UseGas(gasToUse)
+
+	if keyLength != secp256k1CompressedPublicKeyLength && keyLength != secp256k1UncompressedPublicKeyLength {
+		arwen.WithFault(arwen.ErrInvalidPublicKeySize, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	key, err := runtime.MemLoad(keyOffset, keyLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	message, err := runtime.MemLoad(messageOffset, messageLength)
```

Comment:
Consume `DataCopyPerByte` for each byte of `message`, before loading it from memory (`message` has variable length).

Reply:
Done